chronos/futures.nim

```diff
@@ -0,0 +1,369 @@
+#
+#                     Chronos
+#
+#  (c) Copyright 2015 Dominik Picheta
+#  (c) Copyright 2018-2023 Status Research & Development GmbH
+#
+#                Licensed under either of
+#    Apache License, version 2.0, (LICENSE-APACHEv2)
+#                MIT license (LICENSE-MIT)
+
+import std/[strutils, sequtils]
+import stew/base10
+import "."/[config, srcloc]
+export srcloc
+
+when chronosStackTrace:
+  when defined(nimHasStacktracesModule):
+    import system/stacktraces
+  else:
+    const
+      reraisedFromBegin = -10
+      reraisedFromEnd = -100
+
+when chronosStackTrace:
+  type StackTrace = string
+
+type
+  LocationKind* {.pure.} = enum
+    Create
+    Complete
+
+  CallbackFunc* = proc (arg: pointer) {.gcsafe, raises: [Defect].}
+
+  # TODO v4: consider removing this type and just use plain `CallbackFunc`
+  AsyncCallback* = object
+    function*: CallbackFunc
+    udata*: pointer
+
+  FutureState* {.pure.} = enum
+    Pending, Completed, Cancelled, Failed
+
+  FutureBase* = ref object of RootObj ## Untyped future.
+    when defined(chronosv4):
+      location: array[LocationKind, ptr SrcLoc]
+      callbacks: seq[AsyncCallback]
+      cancelcb: CallbackFunc
+      child: FutureBase
+      state: FutureState
+      error: ref CatchableError ## Stored exception
+      mustCancel: bool
+    else:
+      location*: array[LocationKind, ptr SrcLoc]
+      callbacks*: seq[AsyncCallback]
+      cancelcb*: CallbackFunc
+      child*: FutureBase
+      state*: FutureState
+      error*: ref CatchableError ## Stored exception
+      mustCancel*: bool
```

Comment:
There is no difference between types for `chronosv4` and old one, why this distinction is needed?

Reply:
the old ones are public (`*`) - the new ones are all private and accessed only via `internal*` - unfortunately, there's quite a lot of existing code that does `if fut.failed: echo fut.error` - however, it's currently also possible to do `echo fut.error` on a pending future which seems off

Reply:
But ```fut.failed(): echo fut.error()``` no?